bundle/src/main/java/com/adobe/acs/commons/ondeploy/impl/OnDeployExecutorImpl.java

```diff
@@ -0,0 +1,239 @@
+import org.apache.felix.scr.annotations.ReferencePolicy;
+import org.apache.felix.scr.annotations.Service;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.commons.classloader.DynamicClassLoaderManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.Value;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A service that triggers scripts on deployment to an AEM server.
+ * <p>
+ * This class manages scripts so that they only run once (unless the script
+ * fails).  Script execution statuses are stored in the JCR @
+ * /var/acs-commons/on-deploy-scripts-status.
+ * <p>
+ * Scripts are specified via OSGi config, are are run in the order specified.
+ * <p>
+ * NOTE: Since it's always a possibility that
+ * /var/acs-commons/on-deploy-scripts-status will be deleted in the JCR,
+ * scripts should be written defensively in case they are actually run more
+ * than once.  This also covers the scenario where a script is run a second
+ * time after failing the first time.
+ */
+@Component(
+        label = "ACS AEM Commons - On-Deploy Scripts Executor",
+        description = "Developer tool that triggers scripts (specified via an implementation of OnDeployScriptProvider) to execute on deployment.",
+        immediate = true,
+        metatype = true,
+        policy = ConfigurationPolicy.REQUIRE
+)
+@Service
+public class OnDeployExecutorImpl implements OnDeployExecutor {
+    private static final String SCRIPT_DATE_END = "endDate";
+    private static final String SCRIPT_DATE_START = "startDate";
+    private static final String SCRIPT_STATUS = "status";
+    private static final String SCRIPT_STATUS_JCR_FOLDER = "/var/acs-commons/on-deploy-scripts-status";
+    private static final String SCRIPT_STATUS_FAIL = "fail";
+    private static final String SCRIPT_STATUS_RUNNING = "running";
+    private static final String SCRIPT_STATUS_SUCCESS = "success";
+
+    private static final String SERVICE_NAME = "on-deploy-scripts";
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    @Reference
+    private QueryBuilder queryBuilder;
+    @Reference
+    private ResourceResolverFactory resourceResolverFactory;
+    @Reference
+    private DynamicClassLoaderManager dynamicClassLoaderManager;
+    @Reference(name = "scriptProvider", referenceInterface = OnDeployScriptProvider.class, cardinality = ReferenceCardinality.MANDATORY_MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    private List<OnDeployScriptProvider> scriptProviders;
+
+    // TODO: These were needed for injection to work in tests
+    protected void bindQueryBuilder(QueryBuilder queryBuilder) {
```

Comment:
These `bindQueryBuilder`, `bindResourceResolverFactory`, and `bindDynamicClassLoaderManager` functions were needed for unit tests to work, but are not needed in true runtime.  Anyone have a clue if there's just something I need to change in the sling-mock tests to make these `bind` functions unnecessary?

Reply:
You couldnt use Mockito's @Inject?

Reply:
maybe, but that kinda subverts the usage of sling mock, no?  sling mock is successfully seeing the `@Reference` attributes but when it then tries to set them via call to `bind{name}` whereas the real osgi runtime does not require a `bind{name}` function.   At least that's my understanding.  Maybe one of these build plugins we use is actually adding the `bind{name}` functions to the classes automatically at build time?

Reply:
heh...I think I answered my own question...

I took a look at `com.adobe.acs.commons.oak.impl.EnsurePropertyIndex` in the code - no methods for `bindCapabilityHelper` or `bindRepository` to match up with

```
    @Reference
    private AemCapabilityHelper capabilityHelper;

    @Reference
    private SlingRepository repository;
```

However, if I look at the class in the compiled jar, this is what I find:

```
  protected void bindCapabilityHelper(AemCapabilityHelper paramAemCapabilityHelper)
  {
    this.capabilityHelper = paramAemCapabilityHelper;
  }
  
  protected void unbindCapabilityHelper(AemCapabilityHelper paramAemCapabilityHelper)
  {
    if (this.capabilityHelper == paramAemCapabilityHelper) {
      this.capabilityHelper = null;
    }
  }
  
  protected void bindRepository(SlingRepository paramSlingRepository)
  {
    this.repository = paramSlingRepository;
  }
  
  protected void unbindRepository(SlingRepository paramSlingRepository)
  {
    if (this.repository == paramSlingRepository) {
      this.repository = null;
    }
  }
```

So these bind functions must need to be explicit for sling mock if running tests in intellij, because intellij is not running whatever maven plugin is automatically adding these functions.

I guess I'll keep the functions then, since the compile process adds them otherwise anyway.